esp-hal/src/lib.rs

```diff
@@ -572,6 +572,115 @@ mod critical_section_impl {
     }
 }
 
+// The state of a re-entrant lock
+pub(crate) struct LockState {
+    #[cfg(multi_core)]
+    core: portable_atomic::AtomicU8,
+}
+
+impl LockState {
+    pub const fn new() -> Self {
+        Self {
+            #[cfg(multi_core)]
+            core: portable_atomic::AtomicU8::new(0),
+        }
+    }
+
+    #[cfg(multi_core)]
+    fn current_core() -> u8 {
+        // Note: 0 means unlocked.
+        match get_core() {
+            Cpu::ProCpu => 1,
+            Cpu::AppCpu => 2,
+        }
+    }
+}
+
+// This is preferred over critical-section as this allows you to have multiple
+// locks active at the same time rather than using the global mutex that is
+// critical-section.
+#[allow(unused_variables)]
+pub(crate) fn lock<T>(state: &LockState, f: impl FnOnce() -> T) -> T {
+    // In regards to disabling interrupts, we only need to disable
+    // the interrupts that may only eb calling this function.
+
+    #[cfg(not(multi_core))]
+    {
+        // Disabling interrupts is enough on single core chips to ensure mutual
+        // exclusion.
+
+        #[cfg(riscv)]
+        return riscv::interrupt::free(f);
+        #[cfg(xtensa)]
+        return xtensa_lx::interrupt::free(|_| f());
+    }
+
+    #[cfg(multi_core)]
+    {
+        use portable_atomic::Ordering;
+
+        let current_core = LockState::current_core();
+
+        // Check for re-entry
```

Comment:
Since you disable interrupts anyway, I don't think this check is useful except slowing down the common case of non-recursive use. I'd only keep this in release mode and panic if the locked core is the current core, and consider this case a programmer's error.

Reply:
Or similarly to above, we can maybe do this check after we "lost" the CAS check. It'll not be an optimisation, but it'll still allow the inner caller in to the locked code.

Reply:
Already gone haha. I like the idea of no re-entry for now. Also keeps the code a bit simpler.